File: cs/Problems/Day5.cs
namespace AOC2023.Problems;

public class Day5 : IProblem<uint>
{
    public uint Solve(ReadOnlySpan<char> input) => FindLowestSeedLocation(input);

    private uint FindLowestSeedLocation(ReadOnlySpan<char> input)
    {
        Span<Range> linesRange = stackalloc Range[190];
        Span<uint> seedsRange = stackalloc uint[20];

        int currentMap = 0;
        var maps = new List<RangeDiff>[] { [], [], [], [], [], [], [], };

        int lines = input.Split(linesRange, InputReader.NewLine);
        for (int i = 3; i < lines; i++)
        {
            var line = input[linesRange[i]];

            if (line.Contains(':'))
                currentMap++;
            else if (!line.IsEmpty)
                maps[currentMap].Add(RangeDiff.Parse(line));
        }

        uint lowest = uint.MaxValue;

        var seedsString = input[(input.IndexOf(':') + 2)..input.IndexOf(InputReader.NewLine)];
        int parsedSeeds = InputParser.ParseNumbers(seedsRange, seedsString);

        for (int i = 0; i < parsedSeeds; i++)
        {
            uint destination = seedsRange[i];

            foreach (var ranges in maps)
                destination += GetDiffForValueInRange(ranges, destination);

            lowest = Math.Min(lowest, destination);
        }

        return lowest;
    }

    private static uint GetDiffForValueInRange(List<RangeDiff> ranges, uint value)
    {
        foreach (var range in ranges)
            if (value >= range.Start && value <= range.End)
                return range.Diff;

        return 0;
    }

    readonly record struct RangeDiff(uint Start, uint End, uint Diff)
    {
        public static RangeDiff Parse(ReadOnlySpan<char> mapString)
        {
            int separator1 = mapString.IndexOf(' ');
            int separator2 = mapString.LastIndexOf(' ');

            uint destination = uint.Parse(mapString[..separator1]);
            uint source = uint.Parse(mapString[(separator1 + 1)..separator2]);
            uint range = uint.Parse(mapString[(separator2 + 1)..]);

            return new RangeDiff(source, source + range - 1, destination - source);
        }
    }
}

File: ts/problems/day5.ts
import { NEW_LINE } from "../inputReader.js";

const findLowestSeedLocation = (input: string) => findLowestSeed(input.split(NEW_LINE));

const findLowestSeed = (lines: string[]) => {
  // Skip first two lines, then parse the ranges.
  const maps = lines.slice(2).reduce<RangeDiff[][]>((maps, line) => {
    if (line.endsWith(":")) maps.push([]); // Header (:) = new section.
    else if (!!line) maps[maps.length - 1].push(parseRangeDiff(line));
    return maps;
  }, []);

  const seeds = lines[0]
    .substring(lines[0].indexOf(":") + 2)
    .split(" ")
    .map(Number);

  const lowest = seeds.reduce((lowest, seed) => {
    const destination = maps.reduce((dest, ranges) => dest + getDiffForValueInRange(ranges, dest), seed);
    return Math.min(lowest, destination);
  }, Number.MAX_VALUE);

  return lowest;
};

const getDiffForValueInRange = (ranges: RangeDiff[], value: number) => {
  const range = ranges.find((r) => value >= r.start && value <= r.end);
  return range?.diff ?? 0;
};

const parseRangeDiff = (mapString: string): RangeDiff => {
  const [destination, source, range] = mapString.split(" ").map(Number);
  return { start: source, end: source + range - 1, diff: destination - source };
};

type RangeDiff = {
  start: number;
  end: number;
  diff: number;
};

export default findLowestSeedLocation;
